[JENKINS-52818] support @Postconstruct on extensions

ndeloof opened the pull request:
See [JENKINS-52818](https://issues.jenkins-ci.org/browse/JENKINS-52818).

Proposed changes add support for @Postconstruct on Extensions, to offer an homogeneous development model with Stapler's data-bound components. 

This change will also help fixing https://github.com/jenkinsci/configuration-as-code-plugin/issues/370 as Extension's _construction_ and _initialisation_ are then split in two distinct steps.

### Proposed changelog entries

* Extensions can now define @Postconstruct lifecycle methods
* Extensions (including Descriptors) don't need to invoke load() explicitly anymore

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). 
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

ndeloof committed:
add support for JSR250 @PostConstruct

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof committed:
with PostConstruct method no need to explicitely invoke load() anymore

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof commented:
> There are 0 new tests failing, 25 existing failing and 118 skipped.

oleg-nenashev commented on core/src/main/java/hudson/triggers/SCMTrigger.java:
```diff
@@ -347,6 +344,7 @@ public boolean isPollingThreadCountOptionVisible() {
         /**
          * Update the {@link ExecutorService} instance.
          */
+        @PostConstruct
```
So you have two annotations on the same class. How will the invocation order work in such case? `load()` must be always invoked first

ndeloof replied:
not on the same class (this is forbidden by the spec) but on parent and derived class, which is already implemented the same way in spring and stapler by invoking in parent-first order

oleg-nenashev replied:
Then it should to the job

oleg-nenashev commented on core/src/main/java/hudson/ExtensionFinder.java:
```diff
@@ -550,6 +558,29 @@ public Object get() {
+            public <T> void onProvision(ProvisionInvocation<T> provision) {
+                final T instance = provision.provision();
+                if (instance == null) return;
+                List<Method> methods = new LinkedList<>();
```
Sorting order is not clear here. It should be somehow documented so that the approach remains compatible if algorithm changes. Maybe `ordinals` will help, and `load()` must have high ordinal then

oleg-nenashev reviewed (dismissed):
@ndeloof It is a great idea, but IIUC the current implementation introduces some compatibility risks. I would like to discuss them.

The main problem in the implementation is that `load()` call will be invoked twice in all descriptors outside the core. IMO it should be somehow prevented, e.g. by introducing a new `AutoLoadableDescriptor` interface and enabling `load()` only for it. It will also help to prevent compatibility issues if a `Descriptor` implementation does not not actually want  to invoke load() during construction.

ndeloof commented:
I wasn't sure how to prevent `load()` being invoked twice. I thought it wasn't a major issue, just some extra cycles spent at boot. Will check if we can rely on a marker interface for this purpose.

ndeloof committed:
introduce PersistentDescriptor to flag Descriptors
which expect load() to be invoked.

This prevent double execution for plugins which
have been designed on previous version of code and
don’t expect load to be automatically invoked.

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof committed:
document PostConstruct method discovery logic

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

oleg-nenashev commented:
Test failures seem to be related. Maybe `load()` is not invoked in all cases

ndeloof committed:
Force reflective access to non-public @PostConstruct methods

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

ndeloof commented:
I'm stuck here, and so disappointed :'(

ndeloof committed:
check @PostConstruct annotation both on method and on implemented interface with a matching method.

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

oleg-nenashev reviewed (approved):
LGTM since the test pass. @jenkinsci/code-reviewers extra feedback would be helpful

oleg-nenashev commented:
@jenkinsci/code-reviewers LTS cutoff has passed, so I think we can give it a try unless somebody votes against

jglick commented on core/src/main/java/hudson/ExtensionFinder.java:
```diff
@@ -49,20 +52,25 @@
 import net.java.sezpoz.IndexItem;
 import org.kohsuke.accmod.Restricted;
 import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.springframework.util.ClassUtils;
```
Avoid references to Spring if at all possible please.

jglick commented on core/src/main/java/hudson/ExtensionFinder.java:
```diff
@@ -503,6 +511,9 @@ private void resolve(Class c) {
             @Override
             protected void configure() {
+
+                bindListener(Matchers.any(), this);
```
Eek, Guice. My sign to exit PR review immediately and pray.

jglick commented on core/src/main/java/hudson/model/PersistentDescriptor.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Marker interface for Descriptors which use xml persistent data, and as such need to load from disk when instantiated.
+ * <p>
+ * {@link Descriptor#load()} method is annotated as {@link PostConstruct} so it get autmatically invoked after
```
typo

daniel-beck replied:
x2

jglick reviewed (commented):
More Guice magic. Hard to predict what will break. Do not see anything wrong.

daniel-beck commented:
Aren't we on a pretty outdated spring? Seems letting it dig even deeper into Jenkins will make any future update efforts even more painful.

ndeloof commented:
@daniel-beck we can vendor (i.e. copy/paste) Spring's `ClassUtils#getAllInterfaces` method in jenkins codebase if you think it makes sense, anyway I'm not sure this is the hard part into updating spring dependency. I anyway don't expect we can expect any move from this point of view until we get some strong architecture decision to break compatibility, and doing so adopt a complete distinct architecture that will make mostly everything Jenkins obsolete.

@jglick I tend to agree use of Guice in Jenkins is not the best architecture choice (#sarcasms) especially considering it also uses Spring. I hardly understand how such thing did happen, would need to spend some spare cycle to do some git archeology and try to understand the intent behind introducing **two** D.I. frameworks in the same project, and still rely everywhere on `getExtension` lookup method. To be honest, I wish I can introduce in another PR some more Guice magic so we can use `@Inject List<Foo>` and get rid of `Jenkins.getInstance`, but based on your comment I assume you won't volunteer to review it :)

ndeloof committed:
typo

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

oleg-nenashev commented:
It is a bit YOLO, but we have a lot of time till the next LTS baseline. Will revert the change in the worst case

jimklimov commented:
Hi all, just in case: this change seems to have broken at least the Stash Pull Request Builder plugin, as it used a `new` configuration object and expected to have actual server configs there - but for whatever reason, they did not get loaded on their own (I'm not sure how the added annotations are supposed to work).

The hard part of the fix was finding why the plugin complained, and then either adding a `load()` in the new object, or using a `get()` method to instantiate an object -- both worked for me in 2.141 script console. See https://github.com/jenkinsci/stash-pullrequest-builder-plugin/pull/26 for code.

ndeloof commented:
@jimklimov sorry to hear this. Having a quick look you better should rely on `jenkins.model.JenkinsLocationConfiguration#get` to access this extension (which is a singleton). Future of this code might not support concurrent instances within same JVM.

jimklimov commented:
Literally so, instead of `...get()` I used before? (Sorry Java is not my
native language)

I'll give it a shot, if it compiles - will update my PR for Stash Pull
Request Builder ;)

On Thu, Sep 20, 2018, 16:05 Nicolas De loof <notifications@github.com>
wrote:

> @jimklimov <https://github.com/jimklimov> sorry to hear this. Having a
> quick look you better should rely on
> jenkins.model.JenkinsLocationConfiguration#get to access this extension
> (which is a singleton). Future of this code might not support concurrent
> instances within same JVM.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/3570#issuecomment-423195479>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABj5lKVXzRzXk_J84q268oxYFPiXSgSQks5uc6CQgaJpZM4VoZrv>
> .
>

ndeloof commented:
you indeed should replace the erronous `[new JenkinsLocationConfiguration();](https://github.com/jenkinsci/stash-pullrequest-builder-plugin/pull/26/files#diff-90603ac32653a103d3efc564777f63cbR52)` trying to re-create and load this component, but just resolve the singleton instance managed by jenkins, using `JenkinsLocationConfiguration.get()`.

Nothing java specific here, mostly Jenkins specific about how components have a managed lifecycle within Jenkins model.

jimklimov commented:
@ndeloof : thanks to the suggestion, I updated the PR for PR builder :)
If only someone were around to merge it, there are many waiting PRs...
